### DISCUSSION TITLE
VRP: add simple static constraint

### DISCUSSION THREAD

**Discussion by Westlife1002** [google/or-tools#2718] [Routing (and legacy CP) questions] [answered]
What is the correct way to add contraint that does not require cumulative variable. All tutorial examples considers constaints associated with cumulative varable like traveling distance, time and loaded weight. If we want to add a simple static constaint, like big truck cannot visit certain customers because of their loading dock is too small, or items with length exceeding the max limit cannot be assigned to certain types of trucks, how do we add such constaint?

**Comment by Mizux** [comment, marked as answer]
Beside the `RoutingDimension::CumulVar()` you also have:

1. `routing.NextVar(index)` this give you the list of allowed location after this location -> usefull if you want to forbid arc `A -> B` i.e. remove `B` from `NextVar(A)`
2. `routing.VehicleVar(index)` this give you the list of allowed vehicle for this location -> this is what your are looking for here...

ref:
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/routing.h#L1209-L1216

**Reply by Westlife1002** [reply]
many thanks. I will do the experiment. However, can we add the constraint like formulating the MIP model? just like adding the constraint for max distance, max load in the tutorial example.

**Comment by Westlife1002** [comment]
Actually, I am looking for more general way of adding the constraint. the solution above might not be suitable for the second constaint like " items with length exceeding the max limit cannot be assigned to certain types of trucks ". 

Do we have a a way to add the constraint like formulating the standard MIP model? just like adding the constraint for not exceeding the max distance, max load in the tutorial example. thanks

**Reply by Mizux** [reply]
sorry, don't get it.

You can access to the underlying CP solver using `routing.solver()`
then you can use the CP solver api: https://github.com/google/or-tools/blob/stable/ortools/constraint_solver/constraint_solver.h#L1018-L2051

note: Routing is unrelated to linear solver and the MPSolver interface

**Reply by Westlife1002** [reply]
well let's do not make it complex. could you just suggest a simple way of adding below constaint:
If I have many items to pickup by trucks , I need to make sure the length of any item should not exceed the max limit  of a truck. i.e. if the item is longer than a small truck, it should be only assigned to a big truck, or if it is even longer than the big truck, this item is invalid for service and should be abandoned.

Many thanks for your advice.

**Reply by Mizux** [reply]
why not pre processing and using `routing.VehicleVar()` ? i.e. when computing the model instead of allowing any vehicle to visit a location simply filter out...

pseudo code
```python
small_vehicles = [0, 1, 2]
middle_vehicles = [3, 4, 5, 6]
large_vehicles = [7, 8]
all_vehicles = small_vehicles + middle_vehicles + large_vehicles
...
manager = pywrapcp.RoutingIndexManager(..., len(all_vehicles),...)                                           
...
# Small items location
for node in small_item_locations:
  index = manager.NodeToIndex(node)
  routing.VehicleVar(index).SetValues(all_vehicles)

for node in middle_item_locations:
  index = manager.NodeToIndex(node)
  routing.VehicleVar(index).SetValues(middle_vehicles + large_vehicles)

for node in large_item_locations:
  index = manager.NodeToIndex(node)
  routing.VehicleVar(index).SetValues(large_vehicles)
```

And to abandon a visit simply use AddDisjunction() as usual....

**Reply by Westlife1002** [reply]
yes, this approach should work. I will give it a try. However, can't we write the constaint in a way like normally what we did in Gurobi/MP solver, i.e.  if length of order(item) from customer_j <= length of truck_i, truck_assign_ij (assignment of truck_i to this customer_j) <= 1 (truck_i can be possibly assigned to pickup this customer_j's  goods), otherwise, truck_assign_ij = 0 (truck_i not valid to be assigned to service customer_j)

BTW, what is the method routing.VehicleVar() used for? where can I see the explanation of it and other built-in method?
Many thanks

**Reply by Mizux** [reply]
Please take a look at routing.h and the source code...

**Reply by Westlife1002** [reply]
Dear Mizux, I have tried and this approach works ok.

However, I cannot find the explanation for many of the built-in methods in routing.h, e.g.
`routing.VehicleVar(index).SetValues`
where can I find SetValues() and VehicleVar(index) documentation?

**Comment by Westlife1002** [comment]
```python
	                if length < 4.2:
				routing.VehicleVar(index).SetValues(truck_all)
				continue
			if length >= 4.2 and length < 5.2:
				routing.VehicleVar(index).SetValues(truck_5M2+truck_6M8+truck_7M6+truck_9M6+truck_13M)
				continue
			if length >= 5.2 and length < 6.8:
				routing.VehicleVar(index).SetValues(truck_6M8+truck_7M6+truck_9M6+truck_13M)
				continue
			if length >= 6.8 and length < 7.6:
				routing.VehicleVar(index).SetValues(truck_7M6+truck_9M6+truck_13M)
				continue
			if length >= 7.6 and length < 9.6:
				routing.VehicleVar(index).SetValues(truck_9M6+truck_13M)
				continue
			if length >= 9.6 and length < 13:
				routing.VehicleVar(index).SetValues(truck_13M)
				continue
```
how to abandon node if the length >= 13?

**Reply by Mizux** [reply]
a node which is in a disjunction, if dropped, it will have its VehicleVar(index) equal to `-1`

So if you can't remove them in a pre-process, allow these nodes to be dropped **and** use `routing.VehicleVar.SetValue(-1)`

**Reply by Westlife1002** [reply]
yes, thank you
